lib/geocoder_api_client.py

```diff
@@ -0,0 +1,117 @@
+        retry_address_df = address_df.loc[retry_indices]
+        geoids.update(self._get_geoids_with_single_request(retry_address_df))
+        return geoids
+
+    def _get_geoids_with_single_request(self, address_df):
+        """
+        Geocodes the addresses in address_df using a single API request.
+
+        Returns a dataframe containing the geoids (or NaN) indexed to match
+        address_df.
+        """
+
+        with BytesIO() as address_stream:
+            address_df.to_csv(address_stream, header=False,
+                              columns=['address', 'city', 'region', 'postal_code'])
+            raw_response = self._send_request(address_stream)
+        response_df = pd.read_csv(BytesIO(raw_response), header=None, dtype=str,
+                                  index_col=0, engine='python', names=[
+                                  'index', 'input_address', 'match', 'match_type',
+                                  'matched_address', 'coordinates', 'tigerline_id',
+                                  'tigerline_side', 'state_id', 'county_id',
+                                  'tract_id', 'block_id'])
+        geoids = (response_df['state_id'] + response_df['county_id'] +
+                  response_df['tract_id']).rename('geoid')
+
+        if len(geoids) != len(address_df):
+            self.logger.error('Size of input to geocoder ({input_size}) does not match size of output ({output_size})'
+                              .format(input_size=len(address_df), output_size=len(geoids)))
+            raise GeocoderApiClientError('Size of input to geocoder ({input_size}) does not match size of output ({output_size})'
+                                         .format(input_size=len(address_df), output_size=len(geoids)))
+
+        return geoids
+
+    def _send_request(self, address_stream):
+        """
+        Sends a request to the geocoder API.
+
+        Returns a csv string where each line contains information about a
+        single geocoded address.
+        """
+
+        self.logger.debug('Sending batch request to geocoder API')
+        address_stream.seek(0)
+        try:
+            response = self.session.post(
+                os.environ['GEOCODER_API_BASE_URL'],
+                files={'addressFile': NamedTextIOWrapper(
+                    address_stream, name='input_addresses.csv', encoding='utf-8')},
+                params={
+                    'benchmark': os.environ['GEOCODER_API_BENCHMARK'],
+                    'vintage': os.environ['GEOCODER_API_VINTAGE']
+                },
+                timeout=300)
+            return response.content
+        except RequestException as e:
+            self.logger.error(
+                'Failed to retrieve geocoded addresses from API: {}'.format(e))
+            raise GeocoderApiClientError(
+                'Failed to retrieve geocoded addresses from API: {}'.format(e)) from None
+
+
+class NamedTextIOWrapper(TextIOWrapper):
```

Comment:
Really like this solution for the geocoder API named file requirement 👍

Reply:
Thanks!